game-core/src/main/java/games/strategy/triplea/ai/tree/BattleStep.java

```diff
@@ -0,0 +1,1216 @@
+            childUnits.getAliveOrWaitingToDieEnemy(),
+            defending,
+            parameters.data);
+    final Tuple<Integer, Integer> maxAttackAndDiceSides =
+        DiceRoll.getMaxAaAttackAndDiceSides(
+            firingGroup, parameters.getData(), defending, unitPowerAndRollsMap);
+    final int maxRollsAvailable =
+        DiceRoll.getTotalAaAttacks(unitPowerAndRollsMap, validAliveOrInjuredTargets);
+    final int planeHitPoints =
+        (allowMultipleHitsPerUnit
+            ? CasualtyUtil.getTotalHitpointsLeft(validAliveOrInjuredTargets)
+            : validAliveOrInjuredTargets.size());
+    final int totalRolls = Math.min(planeHitPoints, maxRollsAvailable);
+    final Map<Integer, Integer> rollsByDicePower =
+        Map.of(maxAttackAndDiceSides.getFirst(), totalRolls);
+    return RollData.of(totalRolls, rollsByDicePower, maxAttackAndDiceSides.getSecond());
+  }
+
+  private RollData getRegularRollData(
+      final boolean defending,
+      final Collection<Unit> firingGroup,
+      final Collection<Unit> validAliveOrInjuredTargets) {
+    final Map<Integer, Integer> rollsByDicePower =
+        getRegularDiceGrouped(parameters, defending, firingGroup, validAliveOrInjuredTargets);
+    final int totalRolls = rollsByDicePower.values().stream().reduce(0, Integer::sum);
+    return RollData.of(totalRolls, rollsByDicePower, parameters.data.getDiceSides());
+  }
+
+  Map<Integer, Integer> getRegularDiceGrouped(
+      final Parameters parameters,
+      final boolean defending,
+      final Collection<Unit> friendlyUnits,
+      final Collection<Unit> enemyUnits) {
+
+    final Map<Unit, IntegerMap<Unit>> unitSupportPowerMap = new HashMap<>();
+    final Map<Unit, IntegerMap<Unit>> unitSupportRollsMap = new HashMap<>();
+    final Map<Unit, DiceRoll.TotalPowerAndTotalRolls> unitPowerAndRollsMap =
+        DiceRoll.getUnitPowerAndRollsForNormalBattles(
+            friendlyUnits,
+            new ArrayList<>(enemyUnits),
+            friendlyUnits,
+            defending,
+            parameters.data,
+            parameters.location,
+            parameters.territoryEffects,
+            parameters.isAmphibious,
+            parameters.amphibiousLandAttackers,
+            unitSupportPowerMap,
+            unitSupportRollsMap);
+
+    final Map<Integer, Integer> rollsByDicePower = new HashMap<>();
+    for (final Unit unit : friendlyUnits) {
+      final DiceRoll.TotalPowerAndTotalRolls totalPowerAndTotalRolls =
+          unitPowerAndRollsMap.get(unit);
+      final int power = totalPowerAndTotalRolls.getTotalPower();
+      final int rolls = totalPowerAndTotalRolls.getTotalRolls();
+      if (power == 0) {
+        continue;
+      }
+
+      rollsByDicePower.compute(
+          power,
+          (key, prevRolls) -> {
+            if (prevRolls == null) {
+              return rolls;
+            } else {
+              return prevRolls + rolls;
+            }
+          });
```

Comment:
compute is nice and all, but I think using a classical approach makes htis more readable:
```suggestion
      rollsByDicePower.put(power, rollsByDicePower.getOrDefault(power, 0) + rolls);
```

Reply:
I actually like the compute syntax.  I can change it, though if you want.